src/common/src/array/iterator.rs

```diff
@@ -82,6 +82,35 @@ impl<'a> Iterator for ArrayImplIterator<'a> {
     }
 }
 
+pub struct MarkedArayRefIter<'a> {
+    vis: &'a Vis,
+    arr: &'a ArrayImpl,
+    pos: usize,
+}
+
+impl<'a> Iterator for MarkedArayRefIter<'a> {
+    type Item = DatumRef<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.pos < self.arr.len() {
```

Comment:
This would be hard for compiler to optimize :(

Reply:
I guess that we need array-bench :)